api/resolvers/vault.js

```diff
@@ -0,0 +1,75 @@
+import { E_VAULT_KEY_EXISTS, GqlAuthenticationError, GqlInputError } from '@/lib/error'
+
+export default {
+  Query: {
+    getVaultEntry: async (parent, { key }, { me, models }, info) => {
+      if (!me) throw new GqlAuthenticationError()
+      if (!key) throw new GqlInputError('must have key')
+
+      const k = await models.vault.findUnique({
+        where: {
+          key,
+          userId: me.id
+        }
+      })
+      return k
+    },
+    getVaultEntries: async (parent, { keysFilter }, { me, models }, info) => {
+      if (!me) throw new GqlAuthenticationError()
+
+      const entries = await models.vaultEntry.findMany({
+        where: {
+          userId: me.id,
+          key: keysFilter?.length
+            ? {
+                in: keysFilter
+              }
+            : undefined
+        }
+      })
+      return entries
+    }
+  },
+  Mutation: {
+    // atomic vault migration
+    updateVaultKey: async (parent, { entries, hash }, { me, models }) => {
```

Comment:
This migrates the vault, storing all reencrypted entries and the new key hash - atomically.